discovery: Rewrite Python APM instrumentation detection

vitkyrka opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

Rewrite and enable Python APM instrumentation detection.

https://datadoghq.atlassian.net/wiki/spaces/UT/pages/4027154514/

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

https://datadoghq.atlassian.net/browse/USMON-1158

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

This includes three levels of tests: one with a canned maps file, one with real map files but with faked libraries and faked python, and an E2E test with real instrumented and non-instrumented servers without faking.

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

vitkyrka committed:
discovery: Remove noop Ruby detector

vitkyrka committed:
discovery: Pass PID to detectors

In preparation for using it from the Python detector.

vitkyrka committed:
usm: Move test libraries to site-packages/ddtrace

We need some shared libraries in specific paths in order to add tests
for the Python APM instrumentation detection in service discovery.
Using symbolic links does not work.  Since the path name does not matter
for this eBPF SSL tests, rename the path to the one needed by the
upcoming service discovery tests.

vitkyrka committed:
discovery: Rewrite Python detection

Rewrite Python detection based on checking for the ddtrace native
libraries in the processes' maps file.

vitkyrka committed:
discovery: Enable Python instrumentation detection from system-probe

Add tests and allow the rewritten Python instrumentation detection to
run from system-probe.

vitkyrka committed:
discovery: Build packages only on Linux

Due to errors with kernel.HostProc not being supported on Windows.

guyarb commented on pkg/collector/corechecks/servicediscovery/apm/detect.go:
```diff
@@ -80,61 +82,42 @@ func isInjected(envs map[string]string) bool {
 
-func rubyDetector(_ []string, _ map[string]string) Instrumentation {
-	return None
+func getMaps(pid int) ([]byte, error) {
+	mapsPath := kernel.HostProc(strconv.Itoa(pid), "maps")
+	bytes, err := os.ReadFile(mapsPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return bytes, nil
 }
```
In openssl we have a better implementation that scans the file rather than loading it fully into memory

guyarb commented on pkg/collector/corechecks/servicediscovery/apm/detect.go:
```diff
@@ -80,61 +82,42 @@ func isInjected(envs map[string]string) bool {
-	return None
+func getMaps(pid int) ([]byte, error) {
+	mapsPath := kernel.HostProc(strconv.Itoa(pid), "maps")
+	bytes, err := os.ReadFile(mapsPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return bytes, nil
```
```suggestion
	return os.ReadFile(mapsPath)
```

guyarb commented on pkg/collector/corechecks/servicediscovery/apm/detect.go:
```diff
@@ -80,61 +82,42 @@ func isInjected(envs map[string]string) bool {
+// For example:
+// 7aef453fc000-7aef453ff000 rw-p 0004c000 fc:06 7895473  /home/foo/.local/lib/python3.10/site-packages/ddtrace/internal/_encoding.cpython-310-x86_64-linux-gnu.so
+// 7aef45400000-7aef45459000 r--p 00000000 fc:06 7895588  /home/foo/.local/lib/python3.10/site-packages/ddtrace/internal/datadog/profiling/libdd_wrapper.so
+func pythonDetector(pid int, _ []string, _ map[string]string) Instrumentation {
```
nit - I think we should create a struct to be passed to the different detectors.
I assume that overtime we will add more fields and attributes to the signature of the detector as different detectors will need additional (unique?) information.

If you agree, then let's create a jira ticket to follow that

guyarb commented on pkg/collector/corechecks/servicediscovery/apm/detect_nix_test.go:
```diff
@@ -3,12 +3,11 @@
 // Copyright 2016-present Datadog, Inc.
 
-//go:build !windows
+//go:build linux
```
maybe we should rename the file as well? (like remove `_nix`)

vitkyrka committed:
discovery: Do not look for /site-packages/

vitkyrka committed:
discovery: Use scanner for reading maps

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into discovery-apm-python

 Conflicts:
	pkg/collector/corechecks/servicediscovery/service_detector.go
	pkg/collector/corechecks/servicediscovery/service_detector_test.go

vitkyrka committed:
discovery: e2e: Add Python package

Add an __init__.py so that the python server is recognized as being in a
package which leads to a different service name.

vitkyrka committed:
discovery: e2e: Add instrumented Python service

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into discovery-apm-python

vitkyrka commented:
/merge
